posts/2019-01-22-the-5-raisons-detre-of-testing.md

```diff
@@ -0,0 +1,219 @@
+believe tests are important, in no particular order.
+
+## Ensure that the code you've written works
+
+This may be the most straightforward reason to write tests. However, while it's
+valuable to know that whatever you've just written works, it's probably the
+least important reason.
+
+> Testing that code works is the _raison d'être_ of tests!!! How dare you say
+> that it's not important???
+>
+> -- Jimmy from Nebraska
+
+Don't get it twisted Jimmy, I didn't say it was unimportant: the other reasons
+are just _more_ important. Most of the time you'll eventually try the feature
+(open the website, click `Siamese`, watch in awe as your cat appears on the
+screen) after having written the code -- which makes tests redundant. You will
+catch bugs like "trying to open `cat.jpg` instead of `cat.png`" straight away
+anyway.
+
+Running tests as you write the code is mostly a time saver when you develop a
+feature, as you (usually) have a single command to run in order to check that
+you've done things correctly. Also, by writing tests close to your spec, it
+acts as a deadline (once your tests are green, you're done, you can move on to
+something else). Moreover by _first_ writing a (failing) test when tasked with
+fixing a bug you ensure that the bug actually exists. This is typical
+[TDD](https://en.wikipedia.org/wiki/Test-driven_development).
+
+## Ensure that the code you've written works elsewhere
+
+
+<div class="story">
+You wrote your feature and Tommy wrote his part. Feature "ninja" got merged
+into the master branch, after you diligently tested it out by hand. You go home
+that evening, and as you're about to start watching the third season of The
+Wire (for the second time, because you first assumed you didn't need English
+subtitles) you get an angry phone call from Jason, the sales rep working late
+nights because he's demoing the product to people in another timezone. Jason
+explains that he can only get a picture of the Persian cat. Even when he picks
+the Siamese option from the dropdown. The debugging process goes like this:
+
+> You: What OS are you using?
+>
+> Jason: What's an _awe-hess_?
+
+At which point you realize that Jason is running your code on his Windows
+machine. Which, unfortunately enough, cannot tell the difference between
+`cat.png` and `CAT.png`. You sigh, put the remote down and go back to the
+drawing board.
+
+</div>
+
+
+Test suites are particularly important in projects that run in several
+environments. You've written code that works on your machine? Great, but it's
+likely that you forgot to update some config and that the component you just
+wrote will break once it's deployed somewhere else, because you relied on
+something specific to your (highly tuned, very particular) system.
+
+If you have a unit test, you'll make sure it also works on the CI machine(s).
+If you have an integration test for it, you'll make sure it works on the CI
+machine(s) and on the various clusters that a deploy goes through.
```

Comment:
Maybe a build matrix (I believe that's what it's often called in testing) is worth mentioning here, to test different versions in different environments?